### PR TITLE
[SILOptimizer] Alter FSO arg explosion heuristic.

### PR DESCRIPTION
**Pull request by nate-chandler**
The new rule on the basis of which an argument will be exploded is as follows:

(1) An argument with more than 3 live leaves after considering the results of the owned-to-guaranteed transformation will not be exploded.

(2) Rule (1) permitting, an argument all of whose non-trivial leaves are live will be exploded if and only if (a) there is more than one non-trivial leaf and (b) specialization will not result in a thunk.  This rule was beneficial before semantic ARC because it avoided collapsing RC identities.  It is not useful in the face of semantic ARC and should be retired.

(3) Rule (1) permitting, an argument with dead non-trivial leaves will be exploded.

rdar://problem/39957093

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please smoke benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp [hunk 1]
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
...
+  //       leafCount misses cases where leaves are only live until the
+  //       owned-to-guaranteed transform eliminates their liveness.
+
+  unsigned const nontrivialLeafCount =
+      llvm::count_if(allLeaves, [&](SILType type) { return !type.isTrivial(*function); });
+
+  assert(nontrivialLeafCount >= 0 && "Counting non-trivial leaves should result in a non-negative number.");
+
+  // If the argument is completely trivial, it will not be exploded.
+  if (nontrivialLeafCount == 0) {
```
Comment (eeckstein) [inline lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: Again, this condition checks the same as the assert above.

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp [hunk 2]
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
...
+  assert(liveNontrivialLeafCountUpperBound <= nontrivialLeafCount && "There should be no more *live* non-trivial leaves than there are *total* non-trivial leaves.");
+  assert(nontrivialLeafCount <= leafCount && "There should be no more *non-trivial* leaves than there are *total* leaves.");
+
+  // The heuristic max explosion size for a single argument.
+  unsigned const maxExplosionSize = 3;
+
+  // If it is known without taking the owned-to-guaranteed transformation into
+  // account both that exploding will reduce ARC traffic (because an upper bound
+  // for the number of live non-trivial leaves is less thann the non-trivial 
+  // leaf count) and also that the explosion will fit within the heuristic upper
```
Comment (eeckstein) [inline lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: Actually, when I think about it: is this really true, given that we pass most of the arguments as guaranteed by default? I think it mainly depends on the call site. Only if we need a retain-release pair at the call site, eliminating a dead argument will give a benefit.

**Review by eeckstein — COMMENTED** [review]
Nice!
My main concern here is that I'm not sure if exploding arguments for the sake of reducing ARC traffic helps at all, given that we pass most of our arguments as guaranteed. Some of the assumptions in the FSO come from a time where we had the owned-convention as default.
I think it's worth doing some more experiments (in case you didn't do it yet), e.g. compare benchmark results with disabling argument explosion at all, etc.

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp [hunk 3]
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
...
+/// that the total live leaf count should be in 0>..3):
+///
+/// 1. Some of the non-trivial leaves are dead.  Those fields being passed 
+///    results in superfluous ARC traffic.
+///
+/// 2. None of the non-trivial leaves are dead but (a) there is more than one
+///    live non-trivial leaf and (b) specializing will not introduce a thunk.
+///    Breaking apart the non-trivial leaves is beneficial because it imbues
+///    the leaves with different RC identities, enabling the low level ARC 
+///    optimizer to pair retains/releases in an easier way.
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: Yup!  The comment below

```
/// TODO: Eliminate the second case.  That splitting up is only beneficial 
///       before semantic ARC.
```

is about exactly that, eliminating the second rule once we have ownership SIL at this point.  As you say, I'll run the benchmarks without this rule to see how it works.

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: Okay, opened a second PR here to run the benchmarks without this rule: https://github.com/apple/swift/pull/27252 .

**Comment by gottesmm** [comment]
@nate-chandler btw the <= 6 field heuristic used by loadable from address is here in this utility: https://github.com/apple/swift/blob/master/lib/SILOptimizer/Utils/Local.cpp#L1497

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: @eeckstein it is not that simple. What if we want to move that value pass the call site. Consider a dead guaranteed parameter. Without eliminating the argument we will not be able to reduce the lifetime and do other ARC optimizations.

I would put it this way, it only makes sense if we eliminate an ARC "constraint" on the lifetime of the non-trivially typed thing.

**Comment by nate-chandler** [comment]
@swift-ci please test macos

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp [hunk 4]
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
...
+  //       leaving the count lower.
+  unsigned const liveLeafCountUpperBound = argDesc.ProjTree.getLiveLeafCount();
+
+  assert(liveLeafCountUpperBound >= 0 && "Counting live leaves should result in a non-negative number.");
+
+  // If the argument is completely dead, it will not be exploded.
+  //
+  // Explosion makes sense only if some but not all of the leaves are live.  The
+  // dead argument transformation will try to eliminate the argument.
+  if (liveLeafCountUpperBound == 0) {
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: Forget this comment, I didn't see that the assert is >= 0 and not > 0

File: include/swift/SIL/Projection.h [hunk 5]
```diff
@@ -751,13 +752,15 @@ class ProjectionTreeNode {
...
+
+  ArrayRef<unsigned> getChildProjections() const {
+    return llvm::makeArrayRef(ChildProjections);
   }
 
-  llvm::Optional<Projection> &getProjection() { return Proj; }
+  Optional<Projection> &getProjection() { return Proj; }
 
-  llvm::SmallVector<Operand *, 4> getNonProjUsers() const {
+  const SmallVectorImpl<Operand *> &getNonProjUsers() const {
```
Comment (gottesmm) [inline include/swift/SIL/Projection.h]: As a quick aside: If this is a const, better to return an ArrayRef.

**Comment by Catfish-Man** [comment]
I'm slightly concerned the positive impact of this might be overstated a bit in the benchmark results, since it looks like most of the wins are just repeated variations of 2 tests.

File: lib/SILOptimizer/FunctionSignatureTransforms/FunctionSignatureOpts.h [hunk 6]
```diff
@@ -214,6 +225,18 @@ struct FunctionSignatureTransformDescriptor {
...
+  /// Whether specializing the function will result in a thunk with the same 
+  /// signature as the original function that calls through to the specialized
+  /// function.
+  ///
+  /// Such a thunk is necessary if there is (or could be) code that calls the
+  /// function which we are unable to specialize to match the function's
+  /// specialization.
+  bool willSpecializationIntroduceThunk() {
+    return !hasOnlyDirectInModuleCallers ||
+           hasPublicVisibility(OriginalFunction->getLinkage());
```
Comment (gottesmm) [inline lib/SILOptimizer/FunctionSignatureTransforms/FunctionSignatureOpts.h]: I am not sure if hasPublicVisibility is the right check here. Is that a c function? If so, does this properly handle WMO vs non-WMO?

**Comment by nate-chandler** [comment]
@swift-ci please test compiler performance

**Review by gottesmm — COMMENTED** [review]
Quick comments

**Comment by nate-chandler** [comment]
@Catfish-Man That's fair.  Still doing measurements over here and specifically will be looking at the code size impact on more realistic targets including the compatibility suite.

**Comment by nate-chandler** [comment]
```
LLVM.NumLLVMBytesOutput | 1,797,132,560 | 1,797,052,058 | -80,502 | -0.0%
```

from https://ci.swift.org/job/swift-PR-compiler-performance-macOS/539/artifact/comment.md

**Comment by nate-chandler** [comment]
```
LLVM.NumLLVMBytesOutput | 1,535,314,418 | 1,528,530,034 | -6,784,384 | -0.44%
```

File: lib/SILOptimizer/FunctionSignatureTransforms/FunctionSignatureOpts.h (hunk 6 above)
Comment (nate-chandler) [reply lib/SILOptimizer/FunctionSignatureTransforms/FunctionSignatureOpts.h]: Fixed.

File: include/swift/SIL/Projection.h (hunk 5 above)
Comment (nate-chandler) [reply include/swift/SIL/Projection.h]: Fixed.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
The heuristic is now to explode arguments if one of the following sets of conditions hold:

(1) (a) Specializing the function will result in a thunk.  That is, the
        thunk that is generated cannot be inlined everywhere.
    (b) The argument has dead non-trivial leaves.
    (c) The argument has fewer than three live leaves.

(2) (a) Specializing the function will not result in a thunk.  That is,
        the thunk that is generated will be inlined everywhere and
        eliminated as dead code.
    (b) The argument has dead potentially trivial leaves.
    (c) The argument has fewer than six live leaves.

**Comment by nate-chandler** [comment]
@swift-ci please test linux

**Comment by nate-chandler** [comment]
@swift-ci please test compiler performance

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
Debug:
```
LLVM.NumLLVMBytesOutput | 1,797,952,710 | 1,797,869,638 | -83,072 | -0.0%
```

Release:
```
LLVM.NumLLVMBytesOutput | 1,921,660,574 | 1,911,421,822 | -10,238,752 | -0.53%
```

From https://ci.swift.org/job/swift-PR-compiler-performance-macOS/541/artifact/comment.md of https://ci.swift.org/job/swift-PR-compiler-performance-macOS/541/

**Comment by nate-chandler** [comment]
@swift-ci please test macos platform

File: lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp [hunk 7]
```diff
@@ -26,42 +35,216 @@ static llvm::cl::opt<bool> FSODisableArgExplosion(
...
+///
+/// Our main interest here is to expose more opportunities for ARC. This means
+/// that we are not interested in exploding (and partially DCEing) structs in
+/// the following cases:
+///
+/// 1. Completely dead arguments. This is handled by dead argument elimination.
+///
+/// 2. Values that are completely trivial. By splitting these up we create
+///    more register pressure during argument marshalling and do not really add
+///    any advantage. We only eliminate them
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SILOptimizer/FunctionSignatureTransforms/ArgumentExplosionTransform.cpp]: Fixed!

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test compiler performance

**Comment by nate-chandler** [comment]
@swift-ci please test macos platform

**Comment by nate-chandler** [comment]
@swift-ci please clean smoke test macos platform

**Comment by nate-chandler** [comment]
Release:

```
LLVM.NumLLVMBytesOutput | 1,924,598,716 | 1,920,184,346 | -4,414,370 | -0.23%
```

Debug:

```
LLVM.NumLLVMBytesOutput | 1,796,502,774 | 1,796,441,004 | -61,770 | -0.0%
```

Source:

https://ci.swift.org/job/swift-PR-compiler-performance-macOS/544/artifact/

**Comment by nate-chandler** [comment]
@swift-ci please clean test macos platform

**Comment by nate-chandler** [comment]
@swift-ci please clean test macos platform

**Pull request merged**